File: python/hello.py
#!/usr/bin/env python3
"""
Hello World in Python

This script demonstrates various Python features including:
- Basic print statements
- Functions
- Classes
- List comprehensions
- Type hints (Python 3.5+)
- Decorators

To run:
python hello.py
or
python3 hello.py
"""

import sys
import platform
from datetime import datetime
from typing import List, Optional


def print_hello() -> None:
    """Prints a simple hello message."""
    print("Hello from a function!")


def print_hello_to(name: str) -> None:
    """
    Prints a personalized hello message.
    
    Args:
        name: The name to greet
    """
    print(f"Hello, {name}!")


def get_greeting(name: str = "World") -> str:
    """
    Returns a greeting message.
    
    Args:
        name: The name to greet (default: "World")
    
    Returns:
        A greeting string
    """
    return f"Hello, {name}!"


class Greeter:
    """A simple greeter class."""
    
    def __init__(self, greeting: str = "Hello"):
        """
        Initialize the Greeter.
        
        Args:
            greeting: The greeting to use (default: "Hello")
        """
        self._greeting = greeting
    
    def greet(self, name: str = "World") -> str:
        """
        Generate a greeting message.
        
        Args:
            name: The name to greet (default: "World")
        
        Returns:
            A greeting string
        """
        return f"{self._greeting}, {name}!"
    
    @property
    def greeting(self) -> str:
        """Get the current greeting."""
        return self._greeting
    
    @greeting.setter
    def greeting(self, value: str) -> None:
        """Set a new greeting."""
        self._greeting = value


class Person:
    """A person class demonstrating properties and methods."""
    
    def __init__(self, first_name: str, last_name: str):
        """
        Initialize a Person.
        
        Args:
            first_name: The person's first name
            last_name: The person's last name
        """
        self.first_name = first_name
        self.last_name = last_name
    
    def get_full_name(self) -> str:
        """Return the full name with a greeting."""
        return f"Hello, my name is {self.first_name} {self.last_name}!"
    
    def __str__(self) -> str:
        """String representation of the person."""
        return f"{self.first_name} {self.last_name}"
    
    def __repr__(self) -> str:
        """Official string representation of the person."""
        return f"Person('{self.first_name}', '{self.last_name}')"


def demonstrate_comprehensions() -> None:
    """Demonstrate list comprehensions and filtering."""
    greetings = ["Hello", "Hi", "Hey", "Greetings", "Howdy"]
    
    print("Different greetings:")
    for greet in greetings:
        print(f"- {greet}, World!")
    print()
    
    # List comprehension
    long_greetings = [g for g in greetings if len(g) > 3]
    print("Greetings longer than 3 characters:")
    for greet in long_greetings:
        print(f"- {greet}")
    print()
    
    # Using map and lambda
    uppercase_greetings = list(map(lambda x: x.upper(), greetings))
    print("Uppercase greetings:")
    for greet in uppercase_greetings:
        print(f"- {greet}")
    print()


def show_system_info() -> None:
    """Display system information."""
    print("System Information:")
    print(f"Current Date/Time: {datetime.now()}")
    print(f"Python Version: {sys.version}")
    print(f"Platform: {platform.platform()}")
    print(f"System: {platform.system()}")
    print(f"Machine: {platform.machine()}")
    print(f"Processor: {platform.processor()}")


def main() -> None:
    """Main function demonstrating all features."""
    # Simple Hello World
    print("Hello, World!")
    print()
    
    # Hello World with variables
    greeting = "Hello"
    name = "World"
    print(f"{greeting}, {name}!")
    print()
    
    # Using functions
    print_hello()
    print_hello_to("Python Developer")
    print(get_greeting())
    print(get_greeting("Python Programmer"))
    print()
    
    # Using classes
    greeter = Greeter()
    print(greeter.greet())
    print(greeter.greet("Python Expert"))
    
    greeter.greeting = "Hi"
    print(greeter.greet("Friend"))
    print()
    
    # Using Person class
    person = Person("John", "Doe")
    print(person.get_full_name())
    print(f"String representation: {str(person)}")
    print(f"Repr: {repr(person)}")
    print()
    
    # Demonstrate comprehensions
    demonstrate_comprehensions()
    
    # Display system information
    show_system_info()


if __name__ == "__main__":
    main()

File: javascript/hello.js
#!/usr/bin/env node

/**
 * Hello World in JavaScript (Node.js)
 * 
 * This script demonstrates various JavaScript features including:
 * - Basic console output
 * - Functions (regular and arrow)
 * - Classes (ES6+)
 * - Template literals
 * - Array methods (map, filter, forEach)
 * - Promises and async/await
 * 
 * To run:
 * node hello.js
 */

// Simple Hello World
console.log("Hello, World!");
console.log();

// Hello World with variables
const greeting = "Hello";
const name = "World";
console.log(`${greeting}, ${name}!`);
console.log();

// Function declarations
function printHello() {
    console.log("Hello from a function!");
}

function printHelloTo(name) {
    console.log(`Hello, ${name}!`);
}

function getGreeting(name = "World") {
    return `Hello, ${name}!`;
}

// Arrow functions
const getGreetingArrow = (name = "World") => `Hello, ${name}!`;

// Class definition
class Greeter {
    constructor(greeting = "Hello") {
        this._greeting = greeting;
    }
    
    greet(name = "World") {
        return `${this._greeting}, ${name}!`;
    }
    
    get greeting() {
        return this._greeting;
    }
    
    set greeting(value) {
        this._greeting = value;
    }
}

// Person class
class Person {
    constructor(firstName, lastName) {
        this.firstName = firstName;
        this.lastName = lastName;
    }
    
    getFullName() {
        return `Hello, my name is ${this.firstName} ${this.lastName}!`;
    }
    
    toString() {
        return `${this.firstName} ${this.lastName}`;
    }
}

// Async function example
async function asyncGreeting(name = "World", delay = 1000) {
    return new Promise((resolve) => {
        setTimeout(() => {
            resolve(`Hello (after ${delay}ms), ${name}!`);
        }, delay);
    });
}

// Main execution
(async function main() {
    // Using functions
    printHello();
    printHelloTo("JavaScript Developer");
    console.log(getGreeting());
    console.log(getGreeting("JavaScript Programmer"));
    console.log(getGreetingArrow("JS Coder"));
    console.log();
    
    // Using classes
    const greeter = new Greeter();
    console.log(greeter.greet());
    console.log(greeter.greet("JavaScript Expert"));
    
    greeter.greeting = "Hi";
    console.log(greeter.greet("Friend"));
    console.log();
    
    // Using Person class
    const person = new Person("John", "Doe");
    console.log(person.getFullName());
    console.log(`String representation: ${person.toString()}`);
    console.log();
    
    // Working with arrays
    const greetings = ["Hello", "Hi", "Hey", "Greetings", "Howdy"];
    
    console.log("Different greetings:");
    greetings.forEach(greet => {
        console.log(`- ${greet}, World!`);
    });
    console.log();
    
    // Using array methods
    const longGreetings = greetings.filter(g => g.length > 3);
    console.log("Greetings longer than 3 characters:");
    longGreetings.forEach(greet => {
        console.log(`- ${greet}`);
    });
    console.log();
    
    // Using map
    const uppercaseGreetings = greetings.map(g => g.toUpperCase());
    console.log("Uppercase greetings:");
    uppercaseGreetings.forEach(greet => {
        console.log(`- ${greet}`);
    });
    console.log();
    
    // Async/await example
    console.log("Waiting for async greeting...");
    const asyncResult = await asyncGreeting("Async World", 500);
    console.log(asyncResult);
    console.log();
    
    // Display system information
    console.log("System Information:");
    console.log(`Current Date/Time: ${new Date().toString()}`);
    console.log(`Node.js Version: ${process.version}`);
    console.log(`Platform: ${process.platform}`);
    console.log(`Architecture: ${process.arch}`);
    console.log(`Process ID: ${process.pid}`);
    console.log(`Working Directory: ${process.cwd()}`);
    
    // Memory usage
    const memUsage = process.memoryUsage();
    console.log("\nMemory Usage:");
    console.log(`- RSS: ${(memUsage.rss / 1024 / 1024).toFixed(2)} MB`);
    console.log(`- Heap Total: ${(memUsage.heapTotal / 1024 / 1024).toFixed(2)} MB`);
    console.log(`- Heap Used: ${(memUsage.heapUsed / 1024 / 1024).toFixed(2)} MB`);
    console.log(`- External: ${(memUsage.external / 1024 / 1024).toFixed(2)} MB`);
})();
